File: 01 easy/python/uniqueNumberOfOccurrences.py
# 1207. Unique Number of Occurrences
# https://leetcode.com/problems/unique-number-of-occurrences/description/
# Attempts: 1

from typing import List


def uniqueOccurrences(arr: List[int]) -> bool:
    occurrences = {}
    for n in arr:
        if n in occurrences:
            occurrences[n] += 1
        else:
            occurrences[n] = 1

    count = set()
    for occurrence in occurrences.values():
        if occurrence in count:
            return False
        else:
            count.add(occurrence)
    return True


### TEST CASES
print(uniqueOccurrences([1, 2, 2, 1, 1, 3]))
print(uniqueOccurrences([1, 2]))
print(uniqueOccurrences([-3, 0, 1, -3, 1, 1, 1, -3, 10, 0]))

File: 01 easy/rust/uniqueNumberOfOccurrences.rs
// 1207. Unique Number of Occurrences
// https://leetcode.com/problems/unique-number-of-occurrences/description/
// Attempts: 1

use std::collections::HashMap;
use std::collections::HashSet;

fn main() {
    pub fn unique_occurrences(arr: Vec<i32>) -> bool {
        let mut occurrences = HashMap::new();
        for i in arr {
            if occurrences.contains_key(&i) {
                occurrences.insert(i, occurrences.get(&i).unwrap() + 1);
            } else {
                occurrences.insert(i, 1);
            }
        }

        let mut count = HashSet::new();
        for occurrence in occurrences.values() {
            if count.contains(occurrence) {
                return false;
            } else {
                count.insert(occurrence);
            }
        }
        true
    }

    // TEST CASES
    println!("{}", unique_occurrences([1, 2, 2, 1, 1, 3].to_vec()));
    println!("{}", unique_occurrences([1, 2].to_vec()));
    println!(
        "{}",
        unique_occurrences([-3, 0, 1, -3, 1, 1, 1, -3, 10, 0].to_vec())
    );
}
